src/data/finset/gcd.lean

```diff
@@ -0,0 +1,129 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson
+-/
+import data.finset.fold
+import data.multiset.gcd
+
+/-!
+# GCD and LCM operations on finsets
+
+TODO: simplify with a tactic and `data.finset.lattice`
+-/
+
+variables {α β γ : Type*} [comm_cancel_monoid_with_zero α] [nontrivial α] [gcd_monoid α]
+
+namespace finset
+open multiset
+
+/-! ### lcm -/
+section lcm
+
+/-- Least common multiple of a finite set -/
+def lcm (s : finset β) (f : β → α) : α := s.fold gcd_monoid.lcm 1 f
+
+variables {s s₁ s₂ : finset β} {f : β → α}
+
+lemma lcm_def : s.lcm f = (s.1.map f).lcm := rfl
+
+@[simp] lemma lcm_empty : (∅ : finset β).lcm f = 1 :=
+fold_empty
+
+@[simp] lemma lcm_dvd_iff {a : α} : s.lcm f ∣ a ↔ (∀b ∈ s, f b ∣ a) :=
+begin
+  apply iff.trans multiset.lcm_dvd,
+  simp only [multiset.mem_map, and_imp, exists_imp_distrib],
+  exact ⟨λ k b hb, k _ _ hb rfl, λ k a' b hb h, h ▸ k _ hb⟩,
+end
+
+lemma lcm_dvd {a : α} : (∀b ∈ s, f b ∣ a) → s.lcm f ∣ a :=
+lcm_dvd_iff.2
+
+lemma dvd_lcm {b : β} (hb : b ∈ s) : f b ∣ s.lcm f :=
+lcm_dvd_iff.1 (dvd_refl _) _ hb
+
+@[simp] lemma lcm_insert [decidable_eq β] {b : β} :
+  (insert b s : finset β).lcm f = gcd_monoid.lcm (f b) (s.lcm f) :=
```

Comment:
Can you not use `finset.fold_insert_idem` to handle the case splitting here?

Reply:
Not directly. There are two things I could try: either coming up with a new class of almost-idempotent operations, such that `a * a * b = a * b`, or passing to `associates`, where it is actually idempotent.

Reply:
I think I can do it, but that'll involve putting a `gcd_monoid` structure on associates that isn't there yet. Thus I think I'd rather save it for another PR, that will surely come soon.